interfaces/cython/cantera/utils.pyx

```diff
@@ -117,5 +125,209 @@ cdef anyvalue_to_python(string name, CxxAnyValue& v):
 
 cdef anymap_to_dict(CxxAnyMap& m):
     m.applyUnits()
+    if m.isEmpty():
+        return {}
     return {pystr(item.first): anyvalue_to_python(item.first, item.second)
             for item in m.ordered()}
+
+cdef CxxAnyMap dict_to_anymap(data) except *:
+    cdef CxxAnyMap m
+    for k, v in data.items():
+        m[stringify(k)] = python_to_anyvalue(v, k)
+    return m
+
+cdef get_types(item):
+    """ Helper function used by python_to_anyvalue """
+    if not len(item):
+        # Empty list, so no specific type can be inferred
+        return None, 1
+
+    elif isinstance(item, np.ndarray):
+        if isinstance(item.flat[0], numbers.Integral):
+            itype = numbers.Integral
+        elif isinstance(item.flat[0], numbers.Real):
+            itype = numbers.Real
+        else:
+            itype = item.dtype.type
+        return itype, item.ndim
+
+    else:
+        itype = set()
+        for i in item:
+            if isinstance(i, dict):
+                itype.add(dict)
+            elif isinstance(i, list) or isinstance(i , tuple):
+                itype.add(list)
+            elif type(i) == bool:
+                itype.add(bool)  # otherwise bools will get counted as integers
+            elif isinstance(i, numbers.Integral):
+                itype.add(numbers.Integral)
+            elif isinstance(i, numbers.Real):
+                itype.add(numbers.Real)
+            else:
+                itype.add(type(i))
+
+        if itype == {numbers.Real, numbers.Integral}:
+            itype = {numbers.Real}
+
+        if itype == {list}:
+            inner_types = set()
+            ndim_inner = set()
+            for j in item:
+                type_j, ndim_j = get_types(j)
+                inner_types.add(type_j)
+                ndim_inner.add(ndim_j)
+            if len(inner_types) == 1 and len(ndim_inner) == 1:
+                return inner_types.pop(), ndim_inner.pop() + 1
```

Comment:
If `item` is a triply-nested list, will this return early?

Reply:
This will indeed return early, but triply-nested lists are converted to `vector<AnyValue>` with `AnyValue` subsequently converted to `vector<vector<doubl>>`; per (resolved) conversation above, 3D lists are not handled as `vector<vector<vector<double>>>`. There's a precedent for this behavior in the YAML importer, so this is correct. Will leave a note ...